scarb/src/ops/proc_macro_server/connection.rs

```diff
@@ -0,0 +1,98 @@
+use crossbeam_channel::{Receiver, Sender};
+use proc_macro_server_api::{RpcRequest, RpcResponse};
+use std::{
+    io::{BufRead, Write},
+    thread::JoinHandle,
+};
+
+pub struct Connection {
+    pub sender: Sender<RpcResponse>,
+    pub receiver: Receiver<RpcRequest>,
+    io_threads: IoThreads,
+}
+
+impl Connection {
+    pub fn new() -> Self {
+        let (reader_sender, reader_receiver) = crossbeam_channel::bounded(0);
+        let (writer_sender, writer_receiver) = crossbeam_channel::bounded(0);
+
+        let reader = std::thread::spawn(move || {
+            let stdin = std::io::stdin();
+            let mut stdin = stdin.lock();
+
+            let mut line = String::new();
+
+            loop {
+                line.clear();
+
+                if stdin.read_line(&mut line).is_err() {
+                    eprintln!("Error occurred while reading from stdin");
+
+                    break;
+                }
```

Comment:
1. don't you want to build errors here & use anyhow contexts?
2. you're logging to stderr here, so use `tracing` macros